oci/layer/tar_extract.go

```diff
@@ -370,7 +396,13 @@ func (te *tarExtractor) unpackEntry(root string, hdr *tar.Header, r io.Reader) (
 	// "lower" file still exists (so the hard-link would point to the old
 	// inode). It's not clear if such an archive is actually valid though.
 	if !fi.IsDir() || hdr.Typeflag != tar.TypeDir {
-		if err := te.fsEval.RemoveAll(path); err != nil {
+		if fi.Mode()&os.ModeSymlink != 0 && hdr.Typeflag == tar.TypeDir &&
+			te.mapOptions.KeepDirlinks {
```

Comment:
To me, this nested if looks a bit complicated (and is what stopped me from completing my review when you first sent the PR -- because I tried to come up with a more "clear" conditional and didn't get very far).

At the moment, I'm +1 on the design it just needs to be a bit less difficult to understand. I'll take another look and get back to you.

Reply:
Sure, no worries. Perhaps a comment explaining that we want to write through symlinks instead of overwriting them is enough?

Reply:
Ok, I added the comment (back from vacation now). If you merge this, I have the v2 api done on top of this and I can post that too.